From db833edda7abbcff3ff45b43f7686d3916b643fa Mon Sep 17 00:00:00 2001
From: Bogdan-Gabriel Roman <bogdan-gabriel.roman@nxp.com>
Date: Thu, 8 Jul 2021 10:48:39 +0300
Subject: [PATCH 10/17] pci-s32v234: enable msi capability in RC mode

commit db833edda7abbcff3ff45b43f7686d3916b643fa from
https://source.codeaurora.org/external/autobsps32/linux

MSI Capability is not enabled by default in kernel version 5.10 for RC
mode. Having it disabled results in no interrupt being deliverd to the
EP device.

Upstream-Status: Pending

Issue: ALB-7336
Signed-off-by: Bogdan-Gabriel Roman <bogdan-gabriel.roman@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/pci/controller/dwc/pci-s32v234.c | 26 ++++++++++++++++--------
 1 file changed, 18 insertions(+), 8 deletions(-)

diff --git a/drivers/pci/controller/dwc/pci-s32v234.c b/drivers/pci/controller/dwc/pci-s32v234.c
index 71c757bf6712..763ecd50339c 100644
--- a/drivers/pci/controller/dwc/pci-s32v234.c
+++ b/drivers/pci/controller/dwc/pci-s32v234.c
@@ -1195,10 +1195,20 @@ static struct dw_pcie_host_ops s32v234_pcie_host_ops = {
 	.host_init = s32v234_pcie_host_init,
 };
 
-static int __init s32v234_add_pcie_port(struct pcie_port *pp,
+static int __init s32v234_add_pcie_port(struct dw_pcie *pcie,
 			struct platform_device *pdev)
 {
+	struct pcie_port *pp = &(pcie->pp);
 	int ret;
+
+	pp->ops = &s32v234_pcie_host_ops;
+
+	ret = dw_pcie_host_init(pp);
+	if (ret) {
+		dev_err(&pdev->dev, "failed to initialize host\n");
+		return ret;
+	}
+
 #ifdef CONFIG_PCI_MSI
 	pp->msi_irq = platform_get_irq_byname(pdev, "msi");
 	if (pp->msi_irq <= 0) {
@@ -1215,15 +1225,15 @@ static int __init s32v234_add_pcie_port(struct pcie_port *pp,
 	}
 	dev_info(&pdev->dev, "Allocated line %d for interrupt %d",
 		ret, pp->msi_irq);
-#endif
 
-	pp->ops = &s32v234_pcie_host_ops;
+	if (pci_msi_enabled()) {
+		u8 offset = dw_pcie_find_capability(pcie, PCI_CAP_ID_MSI);
+		u16 flags_val = dw_pcie_readw_dbi(pcie, offset + PCI_MSI_FLAGS);
 
-	ret = dw_pcie_host_init(pp);
-	if (ret) {
-		dev_err(&pdev->dev, "failed to initialize host\n");
-		return ret;
+		flags_val |= PCI_MSI_FLAGS_ENABLE;
+		dw_pcie_writew_dbi(pcie, offset + PCI_MSI_FLAGS, flags_val);
 	}
+#endif
 
 	return 0;
 }
@@ -1365,7 +1375,7 @@ static int s32v234_pcie_probe(struct platform_device *pdev)
 	s32v234_pp->soc_revision = s32v234_pcie_get_soc_revision();
 
 	if (!s32v234_pp->is_endpoint) {
-		ret = s32v234_add_pcie_port(pp, pdev);
+		ret = s32v234_add_pcie_port(pcie, pdev);
 		if (ret < 0)
 			return ret;
 	} else {
-- 
2.17.1

